### PR TITLE
Add 'pred-ctx' to query engine s.t. we don't have to rely on globals in Lucene, #1218

### LINKED ISSUE #1218: Allow pred-fns access to node components
For example #1094 needs to maintain a dynamic var, as pred-fns don't have access to the `lucene-store` component. 

This issue to make make more things available to the pred-fns, one extreme is the entire Crux node being available.

An alternative approach, to satisfy the lucene scenario, is that the Lucene store could itself be modeled as an index-store, and it may then benefit from being more exposed where it's needed.

### PR DESCRIPTION
**Pull request by jarohen**
Fixes #1218 - naming RFC.

### CODE REVIEW

File: crux-lucene/test/crux/lucene_test.clj [hunk 1]
```diff
@@ -11,8 +11,12 @@
   (:import org.apache.lucene.analysis.Analyzer
            org.apache.lucene.document.Document
            org.apache.lucene.queryparser.classic.QueryParser
+           [org.apache.lucene.index DirectoryReader]
+           [org.apache.lucene.store Directory]
            [org.apache.lucene.search BooleanClause$Occur BooleanQuery$Builder Query]))
 
+(require 'crux.lucene.multi-field)
```
Comment (jarohen) [inline crux-lucene/test/crux/lucene_test.clj:18]: aside: this test depends on side effects in that namespace

File: crux-lucene/test/crux/lucene_test.clj [hunk 2]
```diff
@@ -157,6 +162,10 @@
     (with-open [db (c/open-db *api*)]
       (t/is (= prior-score (c/q db q))))))
 
+(defn doc-count [{:keys [^Directory directory]}]
```
Comment (jonpither) [inline crux-lucene/test/crux/lucene_test.clj]: May as well make pvt.

File: crux-lucene/test/crux/lucene_test.clj [hunk 3]
```diff
@@ -168,7 +177,8 @@
     (submit+await-tx [[:crux.tx/put {:crux.db/id "ivan" :name "Ivan"}]])
     (submit+await-tx [[:crux.tx/put {:crux.db/id "ivan" :name "Ivan"}]])
 
-    (t/is (= 2 (l/doc-count)))
+    (t/is (= 2 (doc-count (-> @(get-in *api* [:query-engine :!pred-ctx])
```
Comment (jonpither) [inline crux-lucene/test/crux/lucene_test.clj]: if we keep doing this access pattern, could establish an fn in the lucene fixtures ns

**Review by jonpither — APPROVED** [review]
A nice post-xmas present - thanks @jarohen

**Pull request merged**